File: Algorithms/Searching_Algorithms/Binary_Search/Java/Binary_Search.java
// Binary Search Practice/ Implementation
package binarySearch;

import java.util.Scanner;

public class BinarySearchMain {

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    System.out.println("Enter n i.e total number of elements in array");
    int n = sc.nextInt();

    int arr[] = new int[n];
    System.out.println("enter " + n + " numbers in sorted array:");
    for (int i = 0; i < n; i++)
      arr[i] = sc.nextInt();
    System.out.println("Enter a number to find in the array :");
    int num = sc.nextInt();

    sc.close();

    // creating object of class binary search
    BinarySearch obj = new BinarySearch();
    int res = obj.binarySearch(arr, 0, n - 1, num);
    if (res == -1)
      System.out.println("Number not found");
    else
      System.out.println("number found at index : " + res);
  }
}

class BinarySearch {
  int binarySearch(int arr[], int l, int r, int x) {

    if (r >= l) {

      int mid = (l + r) / 2;
      if (arr[mid] == x)
        return mid;
      if (arr[mid] > x) // if given number is less than mid of the array
        return binarySearch(arr, l, mid - 1, x);

      return binarySearch(arr, mid + 1, r, x);
    }
    // the given number is not present in the array
    return -1;
  }
}

File: Algorithms/Searching_Algorithms/Binary_Search/C/Binary_Search.c
// C program to implement Binary Search algorithm
#include <stdio.h> 

//A iterative binary search function.It returns location of target element in given array arr[l..r] if present,otherwise -1
//We are using an array that is sorted in ascending order. 

int binarySearch(int array[],int l,int r, int n, int target) 
{ 
	l=0;
	r=n-1;
	//Implementing Binary Search
	while (l <= r) 
	{ 
		int mid = (l+r)/ 2; 

		// Checking if target element is the middle element 
		if (array[mid] == target) 
			return mid; 

		// If target is greater, ignore left half 
		if (array[mid] < target) 
			l = mid + 1; 

		// If target is smaller, ignore right half 
		else
			r = mid - 1; 
	} 

	// if we reach here, then element was not present and we return -1
	return -1; 
} 

//Driver program to check the above function
int main(void) 
{ 
	int l,r;
	int array[] = {2,9,17,23,28,36,42,59,77,92}; 
	int n = 10; 
	int target = 77 ; 
	int result = binarySearch(array,l,r,n,target); 
	//checking whether 77 is present or not
	(result == -1) ? printf("Element is not present in array"): printf("Element is present at index %d",result); 
return 0; 
} 
